text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,183 @@
+
+TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.
+
+Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time. And it can tolerate fault of TSO no longer than `30s`, to keep time-related metrics such as RPO reasonable.
+
+Besides, on startup, physical part must be initialized by a successful TSO.
+
+##### 2.2.2 Logical Part of Timestamp
+
+Logical part is advanced on every write request.
+
+Moreover, logical part is advanced on *Leader Transfer*. As TiKV a distributed system, every store in TiKV cluster has a instance of timestamp generator. As every key is existed in only one region and only one peer (the leader) in a region can write, it's safe to generate timestamp locally in each store, except for leader transfer.
+
+On leader transfer, the timestamp generated by other store would be bigger than the store where new leader located, which will violate the causal consistency requirement. We solve this by this method: Every peer observes raft message applied, and keep the maximum timestamp of the region (called `Region-Max-TS`). On become leader, this peer advances timestamp generator of store to be bigger than `Region-Max-TS`. 
+
+Besides, other region changes should be carefully handled:
+
+* On region merge: The `Region-Max-TS` of the region is the bigger one of the original two regions.
+
+* On region split: The `Region-Max-TS` of the new region is copied from the original one.
+
+#### 2.3 Resolved Timestamp
+
+**TiKV-CDC** using *Resolved Timestamp* (or `resolved-ts`) to indicate the largest transaction timestamp of the replication task, to help downstream avoid getting partial data of a transaction. Downstream streaming system can also use `resolved-ts` as [*watermark*](https://www.oreilly.com/radar/the-world-beyond-batch-streaming-102/).
+
+As no transaction support, in RawKV scenario we don't need scanning locks in *lock_cf*. But we should notify `resolved-ts` event generation that the minimum timestamp of data that on the way from *proposed* to *applied* (by [`resolver.track_lock`](https://github.com/tikv/tikv/blob/v5.0.4-20211201/components/resolved_ts/src/resolver.rs#L47-L57)).
+
+### 3. Deletion
+
+To make deletions be captured as change data, we turn physical deletion to logical deletion, i.e., set a *deleted* flag, other than physically delete the entry.
+
+The *garbage collection* of deleted data is implemented by setting TTL to 25 hours after, and *GC-ed* by compaction filter. At the same time, start timestamp (i.e. `start-ts`) of **TiKV-CDC** replication task should not be earlier than 24 hours before.
+
+### 4. Encoding
+
+Timestamp & deleted flag are encoded as following:
+
+```
+r{keyspace id}{user key}{MCE Padding}{^timestamp:u64}: {user value}{expire ts}{meta flags}
```

Comment:
should the `keyspace id` part be removed?

Reply:
No. Why it should be removed ?

Reply:
Ah, it's because we've decided not to include keyspace in API V2.

Reply:
Er... OK, so what's the latest proposal of API V2 ?